### PR TITLE
Add support for gradient_predivide_factor and averaging in Horovod backend.

### PR DESCRIPTION
**Pull request by romerojosh**
This PR implements support for `gradient_predivide_factor` in the `DistributedOptimizer`, similar to that available in [APEX DDP](https://nvidia.github.io/apex/parallel.html#apex.parallel.DistributedDataParallel). The purpose of this is to enable splitting the averaging before and after the allreduce, which can be useful in managing the numerical range for mixed precision computations (see #1699). The `gradient_predivide_factor` is applied as follows: 
```
        If op == Average, gradient_predivide_factor splits the averaging
        before and after the sum. Gradients are scaled by
        1.0 / gradient_predivide_factor before the sum and
        gradient_predivide_factor / size after the sum. 
```
To facilitate this, additional arguments, `prescale_factor` and `postscale_factor`, have been added to the basic `hvd.allreduce` functions, enabling the definition of multiplicative factors to scale the tensors before and after the allreduce respectively. For efficiency, the pre and post-scaling is implemented in the Horovod backend on the fused tensor buffer, rather than through framework level operations. For GPU, this required a CUDA kernel implementation to scale the GPU buffer which in turn, required adding compilation of CUDA code to the current build infrastructure. The changes to the build are a more general version of build code changes to enable CUDA compilation proposed in #1760. 

As an additional general benefit from these changes, gradient averaging in the optimizer can now be carried out within the Horovod backend on the fused tensor buffer using the `postscale_factor` argument, rather than on a tensor by tensor basis at the framework level. This should, in general, be more efficient.

Marking the current PR as WIP, but would appreciate feedback to finish up. The implementation is complete, but there are still a few details that need to be worked out (adding support for Keras, if/how to add this to the adasum path), as well as some additional testing.

### CODE REVIEW

File: horovod/tensorflow/__init__.py [hunk 1]
```diff
@@ -66,14 +67,21 @@ def allreduce(tensor, average=None, device_dense='', device_sparse='',
...
         processes.
     """
     op = handle_average_backwards_compatibility(op, average)
-    # Averaging happens in framework code, so translate that to Sum for the actual call
+
     true_op = Sum if op == Average else op
+    if not rocm_built():
+      if (op == Average):
+        # Averaging happens via postscale_factor.
+        postscale_factor /= size()
```
Comment (tgaddair) [inline horovod/tensorflow/__init__.py:83]: For Elastic mode, having the size defined only during graph construction like this is causing problems when the world size scales up or down.  If possible, it would be good if we could move this into the C++ layer to avoid this problem.

One way we could do this: remove the "true op" concept and just pass the actual op enum to C++.  If it's AVERAGE, we can scale the `postscale_factor` dynamically within the TensorFlow custom op using the updated world size.  Then, we can still pass in SUM to the enqueue function so everything else works the same.

What do you think?

Comment (romerojosh) [reply horovod/tensorflow/__init__.py:83]: I see what you are saying. Perhaps we can move the modification of `postscale_factor` into `EnqueueTensorAllreduce` here, if op is average:
https://github.com/horovod/horovod/blob/b94807d20916a6ac5fd85922e593b2f35f62b4e0/horovod/common/operations.cc#L830-L837

But then this also leads to the question of how to handle the `gradient_predivide_factor` as well. For elastic, are we ok if the `postscale_factor` and `prescale_factor` remain constant w.r.t the world size? In general, the `gradient_predivide_factor` is set based the number of workers.

File: horovod/common/controller.cc [hunk 2]
```diff
@@ -449,6 +449,36 @@ Response Controller::ConstructResponse(std::string& name, int joined_size) {
     }
   }
 
+  // If we are doing an allreduce, check that prescaling and postscaling factors
+  // are identical.
```
Comment (tgaddair) [inline horovod/common/controller.cc:453]: For clarity, can you change this to "are identical across ranks"?

File: examples/mxnet_imagenet_resnet50.py [hunk 3]
```diff
@@ -322,7 +324,8 @@ def evaluate(epoch):
     opt = mx.optimizer.create('sgd', **optimizer_params)
 
     # Horovod: create DistributedTrainer, a subclass of gluon.Trainer
-    trainer = hvd.DistributedTrainer(params, opt)
+    trainer = hvd.DistributedTrainer(params, opt,
```
Comment (tgaddair) [inline examples/mxnet_imagenet_resnet50.py:327]: Regarding use in examples: I think we should try to balance presenting the available features to users with not overwhelming them with lots of parameters.  So for some examples (where we only use the defaults) I would suggest keeping them that way, while for the examples that show things like gradient compression and Adasum, I would add this parameter.  Does that sounds reasonable to you?

File: horovod/torch/optimizer.py [hunk 4]
```diff
@@ -113,7 +116,18 @@ def _allreduce_grad_async(self, p):
         tensor = p.grad
         tensor_compressed, ctx = self._compression.compress(tensor)
 
-        handle = allreduce_async_(tensor_compressed, name=name, op=self.op)
+        if self.op == Average and self.gradient_predivide_factor != 1.0:
```
Comment (tgaddair) [inline horovod/torch/optimizer.py:119]: Since `allreduce_async_` is in the public API, I think we should consider how this affects the user's use this function.  Here we're defining an Average op as a Sum op with a post-scale factor scaled by 1 / size.  But if the user were to call `allreduce_async_` directly with `op=Average` and `postscale_factor=s` they would get a different result.

In the implementation of `_allreduce_async`, it seems that the divisor is only size when ROCm is used.  Which would seem to break `op=Average` when calling `allreduce_async_` directly.  Maybe we can pull this logic into `_allreduce_async` to simplify things?

File: setup.py [hunk 5]
```diff
@@ -148,6 +152,15 @@ def get_cpp_flags(build_ext):
 
     raise DistutilsPlatformError(last_err)
 
+def get_nvcc_flags():
+    default_flags = ['--std=c++11', '-O3', '-Xcompiler', '-fPIC']
+    cc_list_env = os.environ.get('HOROVOD_BUILD_CUDA_CC_LIST')
```
Comment (tgaddair) [inline setup.py:157]: Can you add docs about this flag and its defaults to https://github.com/horovod/horovod/blob/master/docs/install.rst#environment-variables ?

File: horovod/common/controller.cc (hunk 2 above)
Comment (romerojosh) [reply horovod/common/controller.cc]: Yep, no problem. Done.

File: horovod/common/controller.cc [hunk 6]
```diff
@@ -449,6 +449,36 @@ Response Controller::ConstructResponse(std::string& name, int joined_size) {
...
+  double prescale_factor;
+  double postscale_factor;
+  if (message_type == Request::ALLREDUCE ||
+      message_type == Request::ADASUM) {
+    prescale_factor = requests[0].prescale_factor();
+    postscale_factor = requests[0].postscale_factor();
+
+    for (unsigned int i = 1; i < requests.size(); ++i) {
+      if (error) {
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/common/controller.cc:462]: Definitely agree. This part of the code has gotten pretty messy.

File: horovod/common/half.cc [hunk 7]
```diff
@@ -39,6 +39,7 @@ bool is_avx_and_f16c() {
 }
 #endif
 
+#if HAVE_MPI
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/common/half.cc:42]: The reason is because `half.h` and `half.cc` used to only be used in MPI compilation paths. I added usage of these files to `collective_operations.h` in order to use functionality of the fp16 specialization for `ScaleBufferCPUImpl`, so this file is now compiled in all cases and requires the ifdef guard of the MPI specific `float16_sum` function.

File: horovod/common/operations.cc [hunk 8]
```diff
@@ -813,22 +829,34 @@ Status EnqueueTensorAllreduce(std::shared_ptr<OpContext> context,
...
+    postscale_factor /= horovod_global.controller->GetSize();
+#else
     LOG(ERROR, horovod_global.controller->GetRank()) << "Enqueuing AVERAGE allreduce is not allowed.";
     return status.Aborted("AVERAGE not allowed.");
+#endif
+  } else if (reduce_op == ReduceOp::ADASUM) {
+#if HAVE_NCCL
+    if (device != CPU_DEVICE_ID) {
+      // Averaging by local size happens via postscale_factor
+      postscale_factor /= horovod_global.controller->GetLocalSize();
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/common/operations.cc:849]: Nice catch. It does have that limitation. Added the additional `!HAVE_ROCM` to the if condition.

File: horovod/tensorflow/__init__.py [hunk 9]
```diff
@@ -40,7 +40,8 @@
 
 
 def allreduce(tensor, average=None, device_dense='', device_sparse='',
-              compression=Compression.none, op=None):
+              compression=Compression.none, op=None,
+              prescale_factor=1.0, postscale_factor=1.0):
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py:53]: Yes. In the typical use case with `gradient_predivide_factor`, `prescale_factor = 1 / gradient_predivide_factor` and `postscale_factor = gradient_predivide_factor / size`.

File: horovod/tensorflow/__init__.py [hunk 10]
```diff
@@ -65,14 +66,22 @@ def allreduce(tensor, average=None, device_dense='', device_sparse='',
...
         processes.
     """
     op = handle_average_backwards_compatibility(op, average)
-    # Averaging happens in framework code, so translate that to Sum for the actual call
-    true_op = Sum if op == Average else op
+
+    average_in_framework = False
+    if rocm_built():
+        # For ROCm, perform averaging at framework level
+        if op == Average or op == Adasum:
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py]: Works for me. Changed it.

File: horovod/tensorflow/__init__.py [hunk 11]
```diff
@@ -202,7 +219,18 @@ def after_create_session(self, session, coord):
...
 def _make_allreduce_grads_fn(name, device_dense, device_sparse,
-                             compression, sparse_as_dense, op):
+                             compression, sparse_as_dense, op, gradient_predivide_factor):
+    if op == Average and gradient_predivide_factor != 1.0:
+        # Perform averaging via pre/postscaling factors.
+        # Split average operation across pre/postscale factors
+        prescale_factor = 1.0 / gradient_predivide_factor
+        postscale_factor = gradient_predivide_factor / size()
+        true_op = Sum
+    else:
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py:246]: The user isn't in direct control of the prescale and postscale factors in this part of the code, since it is within the `DistributedOptimizer`. Their only mechanism to change these here is through `gradient_predivide_factor`. I did add a new exception to ensure that `gradient_predivide_factor` is only used if `op==Average`. 

If a user wants to apply pre- and post-scale factors with other operations (sum, adasum), they can do so using the normal `allreduce` function.

File: horovod/tensorflow/__init__.py [hunk 12]
```diff
@@ -420,7 +451,16 @@ def DistributedOptimizer(optimizer, name=None, use_locking=False, device_dense='
       op:
         The reduction operation to use when combining gradients across
         different ranks.
+      gradient_predivide_factor:
+        If op == Average, gradient_predivide_factor splits the averaging
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py:469]: I added exceptions to disallow cases where `gradient_predivide_factor != 1.0` and `op != Average`.

File: horovod/tensorflow/__init__.py [hunk 13]
```diff
@@ -492,13 +533,22 @@ def DistributedGradientTape(gradtape, device_dense='', device_sparse='',
           op:
             The reduction operation to use when combining gradients across
             different ranks.
+          gradient_predivide_factor:
+            If op == Average, gradient_predivide_factor splits the averaging
+            before and after the sum. Gradients are scaled by
+            1.0 / gradient_predivide_factor before the sum and
+            gradient_predivide_factor / size after the sum.
         """
+        if gradient_predivide_factor != 1.0:
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py:555]: Since I added the new exception for `op != Average` here, I left it as is.

File: horovod/torch/optimizer.py (hunk 4 above)
Comment (romerojosh) [reply horovod/torch/optimizer.py:119]: This is a similar to TF, in that in this section of the code, the user does not have explicit control over the pre- and post-scale factors but only indirect control through `gradient_predivide_factor`. 

For the `allreduce_async_` implementation, they could set `op=Average` and both scale factors and it should work fine (i.e. the tensors will be prescaled before allreduce by `prescale_factor`, postscaled after allreduce by `postscale_factor` and divided by `size` for the average. When `op==Average`, we do not use divisor, but instead apply the additional `size` factor into `postscale_factor` in `EnqueueTensorAllreduce` here: https://github.com/horovod/horovod/blob/b0e51db4e26ea5be153cf3f3489660ebd0e310e3/horovod/common/operations.cc#L840

I'm can't think of a situation where someone would want to mix pre- and post-scale factors with `op==Average` but I do think this is doing what would be expected in that case.

File: setup.py (hunk 5 above)
Comment (romerojosh) [reply setup.py:163]: I added documentation for this flag. Right now, the default compute capability is being detected by NVCC at build time. Do you think we should add a set of default compute capabilities to target a particular set of GPU architectures in all builds?

File: examples/mxnet_imagenet_resnet50.py (hunk 3 above)
Comment (romerojosh) [reply examples/mxnet_imagenet_resnet50.py:327]: I went ahead and remove the `gradient_predivide_factor` argument from some examples to simplify things.

File: horovod/tensorflow/__init__.py [hunk 14]
```diff
@@ -202,7 +219,18 @@ def after_create_session(self, session, coord):
 
 @_cache
 def _make_allreduce_grads_fn(name, device_dense, device_sparse,
-                             compression, sparse_as_dense, op):
+                             compression, sparse_as_dense, op, gradient_predivide_factor):
+    if op == Average and gradient_predivide_factor != 1.0:
```
[the remark this answers is not in the document]
Comment (romerojosh) [reply horovod/tensorflow/__init__.py:240]: If it's alright with you, my preference would be to leave the MXNet code as is (averaging in the C++ code, now for ROCm only) since that is how it was originally implemented and will be removed eventually anyway.

File: setup.py (hunk 5 above)
Comment (tgaddair) [reply setup.py:163]: I think the way you have it now is reasonable.  Thanks for updating the docs!

**Comment by eric-haibin-lin** [comment]
@szhengac FYI
